chain/client/src/test_utils.rs

```diff
@@ -1344,6 +1347,42 @@ impl TestEnv {
         self.process_block(id, block.unwrap(), Provenance::PRODUCED);
     }
 
+    /// Pause processing of the given block, which means that the background
+    /// thread which applies the chunks on the block will get blocked until
+    /// `resume_block_processing` is called.
+    ///
+    /// Note that you must call `resume_block_processing` at some later point to
+    /// unstuck the block.
+    ///
+    /// Implementation is rather crude and just hijacks our logging
+    /// infrastructure. Hopefully this is good enough, but, if it isn't, we can
+    /// add something more robust.
+    pub fn pause_block_processing(&mut self, capture: &mut TracingCapture, block: &CryptoHash) {
```

Comment:
@mzhangmzz especially curious if you think there's a better way to achieve this! This is the second test where we need to coax async block processing to (not) happen at a particular time.

Reply:
Maybe we can modify the existing `wait_for_all_blocks` and `wait_for_block` methods in the BlocksInProcessing struct? Current tests call these methods to finish the block processing for all blocks. We can add another field `paused_blocks` there and exclude the block in `paused_blocks` when we call these wait for blocks methods.

Reply:
:thinking:, I think that won't work: this'll prevent us from waiting for those blocks, but it won't prevent the blocks from actually getting done. This line is the key: 

https://github.com/near/nearcore/blob/7efd76cd6381f59e4589a6b0fd6f6c3d65492e03/chain/chain/src/chain.rs#L1579

Here, we'll process ready blocks, and we esentially want to force some blocks to stay in that channel. We *could* do that, but that'll I think require modification of production code, and not only the testonly stuff.

Reply:
You are totally right! Sorry I didn't think very carefully.

Reply:
I can't think of a better way. Another way I can think of is add yet another callback before we call `sc.send(block_hash.clone(), res))` to intercept the process there, but then that adds yet another callback and we already have two callback methods in the `schedule_apply_chunks` code.